activation/activation.go

```diff
@@ -106,14 +105,15 @@ type Builder struct {
 	postSetupProvider PostSetupProvider
 	challenge         *types.NIPostChallenge
 	initialPost       *types.Post
+	commitmentAtx     *types.ATXID
```

Comment:
We already have ways of caching values in our application that are concurrency safe (e.g. cachedDB), I would prefer using / extending those over implementing another solution.

The suggested change would probably work, but complicates testing a bit. I don't think accessing `commitmentAtx` directly within the `activation` package is an issue. It behaves currently the same as other fields of the builder some of which also never change during the lifetime of it.

I intentionally did not include any bigger refactorings to limit the size of this PR; the only bigger re-writes I did was re-implementing tests since fixing the existing ones would have been more effort than just implementing all tests anew.

Reply:
I'm fine with using any existing solution. I didn't know about them. 

>The suggested change would probably work, but complicates testing a bit. 

How would it complicate testing?

IMHO it would actually simplify the code (by hiding caching code into a specialized container) and make it error-resistant (by making improper usage impossible). I consider using the `commitmentAtx` field directly improper as it would bring a non-obvious dependency where the getter must be called first (to fill the variable). IMO it's better to write code that is impossible to misuse rather than hope that nobody will use it wrongly in the future. Do you see any cons?

>I intentionally did not include any bigger refactorings to limit the size of this PR

It's not a refactor, that is a totally new piece of code.